### ISSUE TITLE
Enqueueing of a new job inside `WATCH`

### ISSUE THREAD

**Issue by ta2-1** [nvie/rq#505]
Former title was: "Changing `_dependency_id` private attribute of a `Job`".

Let me describe the question. I'm trying to solve the following task:
There are files `A`, `B`, `C`. Data from these files can be loaded to a DB. Suppose we have `job A`, `job B`, `job C` for that. We can't run more than one and the same job in parallel because of  possible race conditions. But different files can be processed in parallel with multiple workers.

I hoped it’s possible to do so with RQ.

My solution is:
- keep `id` of last enqueued `job X`;
- every `job X` can be enqueued or added as a dependency to the last enqueued `job X`.

For instance if no rqworkers are running and these jobs `A`, `B`, `C`, `B1`, `A1`, `C1`, `B2` are to be executed then we get such result:
- default queue: `A` → `B` → `C`
- dependents: `B1(depends_on=B)`, `A1(depends_on=A)`, `C1(depends_on=C)`, `B2(depends_on=B)`

(There is an optimization for the initial task: we can skip adding of a new job (even as a dependency) if there is the same non-started job in the queue. But I leave it aside.)

How can it be implemented with current RQ API?

``` python
last_job_id = queue.connection.get('last_job_X_id')  # (1)
queue.enqueue_call(func, depends_on=last_job_id)  # (2)
```

But we get two jobs running in parallel if another `job X` is added between (1) and (2).

How to enqueue a new job with dependency watching `last_job_X_id`?
Finally this function can be used for adding a new job but it’s necessary to change `_dependency_id` job attribute and the code was copy-pasted from existing mostly.

``` python
def create_new_job(func, *args):
    queue = get_queue('default')
    job = queue.job_class.create(func, args, connection=queue.connection)

    def do_enqueue_job(queue, job, pipe):
        queue.connection.sadd(queue.redis_queues_keys, queue.key)
        job.set_status(JobStatus.QUEUED, pipeline=pipe)
        job.origin = queue.name
        job.enqueued_at = utcnow()
        if job.timeout is None:
            job.timeout = queue.DEFAULT_TIMEOUT
        job.save(pipeline=pipe)

    with queue.connection.pipeline() as pipe:
        while True:
            try:
                pipe.watch('last_job_X_id')
                last_job_id = queue.connection.get('last_job_X_id')
                pipe.watch(queue.job_class.key_for(last_job_id))
                depends_on = Job(id=last_job_id, connection=queue.connection)
                depends_on_status = depends_on.get_status()
                pipe.multi()
                pipe.set('last_job_X_id', job.id)

                if depends_on_status != JobStatus.FINISHED:
                    # add job as a dependent
                    # HACKISH: this relies on modifying an RQ Job's internal attribute,
                    # handle with care, and especially double-check if the attribute
                    # is always present in newer versions of RQ
                    job._dependency_id = last_job_id
                    job.set_status(JobStatus.DEFERRED)
                    job.register_dependency(pipeline=pipe)
                    job.save(pipeline=pipe)
                    pipe.execute()
                    return job

                do_enqueue_job(queue, job, pipe)
                pipe.execute()
                break
            except WatchError:
                continue

    queue.push_job_id(job.id)
```

Is there better solution? And can you add changing `_dependency_id` attribute to API?

Finally I realized that it can be implemented without changing a private attribute but with creation a job every time before `pipe.execute()` inside the infinite while-loop.

``` python
def create_new_job(func, *args):
    queue = get_queue('default')
    job = queue.job_class.create(func, args, connection=queue.connection)
    job_id = job.id

    def do_enqueue_job(queue, job, pipe):
        queue.connection.sadd(queue.redis_queues_keys, queue.key)
        job.set_status(JobStatus.QUEUED, pipeline=pipe)
        job.origin = queue.name
        job.enqueued_at = utcnow()
        if job.timeout is None:
            job.timeout = queue.DEFAULT_TIMEOUT
        job.save(pipeline=pipe)

    with queue.connection.pipeline() as pipe:
        while True:
            try:
                pipe.watch('last_job_X_id')
                last_job_id = queue.connection.get('last_job_X_id')
                pipe.watch(queue.job_class.key_for(last_job_id))
                depends_on = Job(id=last_job_id, connection=queue.connection)
                depends_on_status = depends_on.get_status()
                pipe.multi()
                pipe.set('last_job_X_id', job.id)

                if depends_on_status != JobStatus.FINISHED:
                    # add job as a dependent
                    job = queue.job_class.create(func, args, id=job_id,
                                                 depends_on=last_job_id,
                                                 connection=queue.connection,
                                                 status=JobStatus.DEFERRED)
                    job.register_dependency(pipeline=pipe)
                    job.save(pipeline=pipe)
                    pipe.execute()
                    return job

                do_enqueue_job(queue, job, pipe)
                pipe.execute()
                break
            except WatchError:
                continue

    queue.push_job_id(job.id)
```

@selwin, @nvie could you please share your thoughts and impressions?

**Comment by selwin** [comment]
To answer your question:
```
For instance if no rqworkers are running and these jobs A, B, C, B1, A1, C1, B2 are to be executed then we get such result:

default queue: A → B → C
dependents: B1(depends_on=B), A1(depends_on=A), C1(depends_on=C), B2(depends_on=B)"
```

I would simply do this:
```python
job_A = queue.enqueue(first_job, args=[a])
job_A1 = queue.enqueue(second_job, args=[a], depends_on=job_A)

job_B = queue.enqueue(first_job, args=[b])
job_B1 = queue.enqueue(second_job, args=[b], depends_on=job_B)

job_C = queue.enqueue(first_job, args=[c])
job_C1 = queue.enqueue(second_job, args=[c], depends_on=job_C)
```

This will allow jobs A, B and C to be run in parallel by multiple workers and at the same time respect their dependencies.

**Comment by ta2-1** [comment]
@selwin thanks for your answer.
I would do the same if I could. I missed one important moment when I tried to describe my question as simple as possible. Jobs can be created in different processes. So I don't have an access to `job_A` object when `job_A1` should be enqueued.

**Comment by selwin** [comment]
Can you modify `first_job` to enqueue `second_job` when it completes? Something like this:

```python
def first_job(arg):
    # ...
    queue.enqueue(second_job, args=[arg])
```

**Comment by ta2-1** [comment]
@selwin thank you for your help. But as I mentioned before `second_job` can be run from another process. So there is possibility to enqueue `second_job` from the `first_job` unfortunately. Actually I use this approach for child-parent jobs.

**Comment by selwin** [comment]
This is fixed in https://github.com/nvie/rq/pull/742
